src/aliceVision/image/imageCaching.hpp

```diff
@@ -0,0 +1,408 @@
+    /// thus ensuring that only one of the shared pointers is non-null
+    CacheValue();
+
+public:
+    /**
+     * @brief Template method to get a shared pointer to the image with pixel type given as template argument.
+     * @note At most one of the generated methods will provide a non-null pointer.
+     * @return shared pointer to an image with the pixel type given as template argument
+     */
+    template<typename TPix>
+    std::shared_ptr<Image<TPix>> get();
+
+    /**
+     * @brief Count the number of usages of the wrapped shared pointer.
+     * @return the use_count of the wrapped shared pointer if there is one, otherwise 0
+     */
+    int useCount() const;
+
+    /**
+     * @brief Retrieve the memory size (in bytes) of the wrapped image.
+     * @return the memory size of the wrapped image if there is one, otherwise 0
+     */
+    unsigned long long int memorySize() const;
+
+private:
+    std::shared_ptr<Image<unsigned char>> imgUChar;
+    std::shared_ptr<Image<float>> imgFloat;
+    std::shared_ptr<Image<RGBColor>> imgRGB;
+    std::shared_ptr<Image<RGBfColor>> imgRGBf;
+    std::shared_ptr<Image<RGBAColor>> imgRGBA;
+    std::shared_ptr<Image<RGBAfColor>> imgRGBAf;
+    
+};
+
+template<>
+inline std::shared_ptr<Image<unsigned char>> CacheValue::get<unsigned char>() { return imgUChar; }
+
+template<>
+inline std::shared_ptr<Image<float>> CacheValue::get<float>() { return imgFloat; }
+
+template<>
+inline std::shared_ptr<Image<RGBColor>> CacheValue::get<RGBColor>() { return imgRGB; }
+
+template<>
+inline std::shared_ptr<Image<RGBfColor>> CacheValue::get<RGBfColor>() { return imgRGBf; }
+
+template<>
+inline std::shared_ptr<Image<RGBAColor>> CacheValue::get<RGBAColor>() { return imgRGBA; }
+
+template<>
+inline std::shared_ptr<Image<RGBAfColor>> CacheValue::get<RGBAfColor>() { return imgRGBAf; }
+
+
+/**
+ * @brief A class for retrieving images from disk (optionally downscaled) that implements a caching mechanism.
+ * 
+ * When creating an instance of this class, one must provide two memory size limits: 
+ * - a capacity: the amount of space dedicated to storing images for later use
+ * - a maximal size: the amount of space that cannot be exceeded.
+ * 
+ * When one attempts to retrieve an image through this cache, the following policy applies: 
+ * 1. if the image is already in the cache, return it
+ * 2. if the image fits in the capacity, load it, store it in the cache and return it
+ * 3. find the Leat-Recently-Used image in the cache that is not used externally and which size is bigger than the missing capacity, 
+ * if such an image exists remove it and apply step 2
+ * 4. remove all images that are not used externally from Least- to Most-Recently-Used until the capacity is big enough to fit the image
+ * or until there is nothing to remove
+ * 5. if the image fits in the maximal size, load it, store it and return it
+ * 6. the image is too big for the cache, throw an error.
+ */
+class ImageCache 
```

Comment:
Use the same case for the file: "ImageCache.hpp". So it's clear that there is a corresponding class in this file.